### ISSUE TITLE
Improvements to the retry logic for mutation.

### ISSUE THREAD

**Issue by MikhailBurdukov** [ClickHouse/ClickHouse#57089]
**Use case**

If mutation query is incorrect or can't be performed right now then it will hang in infinite loop with high resources usage:
Example (from #55946): 
```
CREATE TABLE main (id Int8) ENGINE=MergeTree() ORDER BY id;
INSERT INTO main SELECT * FROM system.numbers LIMIT 2;

ALTER TABLE main DELETE WHERE id IN (SELECT id FROM nonexistent);
``` 
BTW: the number of log records is increasing rapidly with:

```
2023.11.22 07:38:54.030832 [ 1198979 ] {17605073-9555-4773-b45c-438dbcccfcca::all_1_1_0_2} <Error> virtual bool DB::MutatePlainMergeTreeTask::executeStep(): Code: 60. DB::Exception: Table default.nonexistent does not exist: While processing id IN ((SELECT id FROM default.nonexistent) AS _subquery642). (UNKNOWN_TABLE), Stack trace (when copying this message, always include the lines below):
```

**Describe the solution you'd like**

The good way solution that I can see:
 1.  Have a  backoff logic (e.g. exponential). Will provide an ability for reduced CPU usage, memory usage and log file sizes.
 2.  Have a logic to automatic mutation canceling.
 
There are several difficulties to contemplate:

- Backoff
  To manage the time until the next mutation retry, we need the time of the last failure and already made attempts count. The `system.mutations` can be extended to keep these values for example. 
 But what if table is replicated then need to synchronize it somehow. What to do if some replicas lag behind? Should we wait until all replicas to complete the mutation? 
 Also in that case we will have a list with last failure times. Which one should be a basis? What if the replicas will have different error codes? 

- Canceling
Better to check this issue with full discussion(#55777). Summary we will need a transactions to canceling mutation.   

The exponential backoff is good starting point I think.
I believe this is a debatable topic and  require more ideas then I have.

**Additional context**
Related issues:
#36987
#55777
#55946

**Comment by MikhailBurdukov** [comment]
@myrrc

**Comment by myrrc** [comment]
Just to add up some more detail: the more-or-less good solution to the backoff issue is having virtual columns on each replica on `system.mutations` table with last error code (not error message), start time, and restarts count.
They wouldn't be visible to user but would be used inside CH. Then we have no issues with inconsistent `system.mutations` state and no need to synchronize it.

**Comment by Alex-Cheng** [comment]
With backoff logic solution, the failed mutaion will infinitely retry but in an exponentially increasing interval. Is my understanding right? In this way, the mutation always resides in the CH server and we must manually cancel/delete them. It sounds good but requires manual maintenance operations(BTW I have no problem for it).

**Comment by MikhailBurdukov** [comment]
Yes, only one restriction is the max time of mutation postponing.
>With backoff logic solution, the failed mutaion will infinitely retry but in an exponentially increasing interval. Is my understanding right?

**Comment by MikhailBurdukov** [comment]
Yes, only one restriction is the max time of mutation postponing.
>With backoff logic solution, the failed mutaion will infinitely retry but in an exponentially increasing interval. Is my understanding right?

**Comment by Alex-Cheng** [comment]
> Yes, only one restriction is the max time of mutation postponing.
> 
> > With backoff logic solution, the failed mutaion will infinitely retry but in an exponentially increasing interval. Is my understanding right?

Is there any setting for specifying "the max time of mutation postponing"?

**Comment by MikhailBurdukov** [comment]
`max_postpone_time_for_failed_mutations_ms` MergeTree setting

**Comment by Alex-Cheng** [comment]
> `max_postpone_time_for_failed_mutations_ms` MergeTree setting

Cool. I find it in the new version of ClickHouse.

**Comment by xoelop** [comment]
If I understand correctly, even the `kill` command doesn't kill mutations (as I mention [here](https://github.com/ClickHouse/ClickHouse/issues/36987#issuecomment-2669133879) - the number of `parts_to_do` kept going down even after running it) - am I right? 

Is there a way to effectively kill a mutation?

Thanks!

**Comment by MikhailBurdukov** [comment]
@xoelop  Seems like your case close to the  scenario from: https://github.com/ClickHouse/ClickHouse/issues/55777#issuecomment-1775272828 

As far as I know there wasn't any big changes in the logic (maybe I'm wrong, double-checking required). 
So canceling the mutation might leads to merges problems. if I were you, I would increase RAM and memory limit, looks safer.

> even the kill command doesn't kill mutations (as I mention in this issue: https://github.com/ClickHouse/ClickHouse/issues/36987#issuecomment-2669133879) - the number of parts_to_do kept going down even after running it) - am I right?

Suggesting `kill` and `parts_to_do kept going down` events are not related. Mutation will continue to be applied to parts which not too big, event without `kill`.
